### PR TITLE
Update Compositions using Palindrom  flow

### PR DESCRIPTION
**Pull request by tomalec**
https://github.com/StarcounterApps/CompositionProvider/issues/12

### CODE REVIEW

**Comment by warpech** [comment]
Did you have success with any of the apps? I tried running People + SignIn + Images:

```js
document.querySelector('puppet-client').obj.CurrentPage.CompositionProvider.Composition$ = `<h1>Hey there</h1>${document.querySelector('puppet-client').obj.CurrentPage.CompositionProvider.Composition$}`
```

And clicked, but the changes seem no to take any effect:

2017-07-12 17_13_49-people

When I set a breakpoint inside https://github.com/StarcounterApps/CompositionProvider/pull/13/files#diff-68f2e8bee68947e75bc1eec116f6b2c1R9 it does not get triggered.

**Comment by tomalec** [comment]
I tried it with Products + Barcodes + CompositionProvider + CompositionEditor.

I'll check with your setup

**Comment by tomalec** [comment]
Which version of Signin are you using? current develop does not build for me.

**Comment by tomalec** [comment]
Ok that was lack of clean in `build.bat` do you think we could add it?

**Comment by tomalec** [comment]
Checked with People + SignIn + Images + CompositionProvider, and it works
composition

**Comment by warpech** [comment]
> Ok that was lack of clean in `build.bat` do you think we could add it?

See: https://github.com/Starcounter/Home/issues/86

I will try testing again on a clean database.

**Comment by warpech** [comment]
Finally got this to work! Actually I needed to clean CompostionProvider solution :)

File: src/CompositionProvider/viewmodels/LayoutInfo.json.cs [hunk 1]
```diff
@@ -4,5 +4,19 @@ namespace CompositionProvider
...
+            if (setup == null)
+            {
+                setup = new HTMLComposition()
+                {
+                    Key = PartialId,
+                    Version = ""
+                };
+            }
+            setup.Value = input.Value;
+            Transaction.Commit();
```
Comment (warpech) [inline src/CompositionProvider/viewmodels/LayoutInfo.json.cs]: There is an exception if you are not in scope of an already created transaction. For example try this with CssVariablesManager (list page) + CompostionProvider. You will get:

```
Starcounter.DbException: ScErrReadOnlyTransaction (SCERR4093): The transaction is readonly and cannot be changed to write-mode.
Version: 2.3.1.6682.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4093.
   at Starcounter.Internal.DbState.Insert(UInt16 tableId, UInt64& oid, UInt64& address) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter\Internal\DbState.cs:line 126
   at Starcounter.HTMLComposition..ctor(UInt16 tableId, TypeBinding typeBinding, Initialized dummy)
   at Starcounter.HTMLComposition..ctor()
   at CompositionProvider.LayoutInfo.Handle(Composition input) in c:\www\CompositionProvider\src\CompositionProvider\viewmodels\LayoutInfo.json.cs:line 12
   at CompositionProvider.LayoutInfo.JsonByExample.Schema.<>c.<.ctor>b__0_12(Json pup, Input`1 input)
   at Starcounter.Templates.Property`1.InvokeHandler(Json parent, Input`1 input) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:line 267
   at Starcounter.Templates.Property`1.ProcessInput(Json parent, T value) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:line 289
   at Starcounter.XSON.DefaultPatchHandler.<>c__DisplayClass0_0.<Handle>b__0() in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:line 50
   at Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Class Json\JsonExtension.cs:line 67
   at Starcounter.XSON.DefaultPatchHandler.Handle(Json root, JsonPatchOperation patchOp, JsonPointer pointer, String value) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:line 38
   at Starcounter.XSON.JsonPatch.HandleOnePatch(Json root, JsonPatchOperation op, JsonPointer ptr, String value, Int32 patchCount, ChangeLog changeLog, Object sourceJson, Boolean strictRejection) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:line 537
   at Starcounter.XSON.JsonPatch.Apply(Json root, JsonTextReader reader, Boolean strictPatchRejection, Object sourceJson, Int32& patchCount) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:line 445
   at Starcounter.XSON.JsonPatch.Apply(Json root, String patch, Boolean strictPatchRejection, Int32& patchCount) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:line 367
   at Starcounter.Internal.PuppetRestHandler.HandleWebSocketJson(String bs, WebSocket ws) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\PuppetRestHandler.cs:line 62
   at Starcounter.Internal.AllWsGroups.RunHandler(UInt16 managedHandlerId, UInt32 groupId, WebSocket ws) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter.Internal\Http\WsManager.cs:line 289
   at Starcounter.GatewayHandlers.HandleWebSocket(UInt16 managedHandlerId, Byte* rawChunk, BMX_TASK_INFO* taskInfo, Boolean* isHandled) in C:\TeamCity\BuildAgent\work\sc-6135\Level1\src\Starcounter\Apps\GatewayHandlers.cs:line 793
ErrorCode=4093
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4093
```

2017-07-13 11_57_54-cssvariablesmanager

**Comment by tomalec** [comment]
Thanks, is it better now?

**Pull request merged**